### PR TITLE
Tune unparsing of hex strings in QPDF_String::unparse

### PR DESCRIPTION
**Pull request by m-holger**
I see an improvement of approx 2%.

### CODE REVIEW

**Review by jberkenbilt — CHANGES_REQUESTED** [review]
I guess it's probably worth duplicating logic here and in QUtil::hex_encode. I bet QUtil::hex_encode would benefit from a similar rewrite, though it's called less often. I'm betting the performance gains would be higher for encrypted files which tend to have a lot more hex strings.

File: libqpdf/QPDF_String.cc [hunk 1]
```diff
@@ -113,7 +113,14 @@ QPDF_String::unparse(bool force_binary)
     bool use_hexstring = force_binary || useHexString();
     std::string result;
     if (use_hexstring) {
-        result += "<" + QUtil::hex_encode(this->val) + ">";
+        auto const hexchars = "0123456789abcdef";
```
[the remark this answers is not in the document]
Comment (m-holger) [reply libqpdf/QPDF_String.cc:116]: yes

File: libqpdf/QPDF_String.cc [hunk 2]
```diff
@@ -113,7 +113,14 @@ QPDF_String::unparse(bool force_binary)
...
+        result.reserve(2 * this->val.length() + 3);
+        result += '<';
+        for (const char c: this->val) {
+            result += hexchars[(c & 0xf0) >> 4];
```
[the remark this answers is not in the document]
Comment (m-holger) [reply libqpdf/QPDF_String.cc:120]: The & 0xf0 eliminates the sign bit and ensures "negative" characters are shifted correctly.

[the remark this answers is not in the document]
Comment (jberkenbilt) [reply libqpdf/QPDF_String.cc:120]: Yeah, you're right. I had it my head that there were an infinite number of `0`s standing off to the left, but it turns out that, for negative numbers, there are an infinite number of `1`s.

[the remark this answers is not in the document]
Comment (m-holger) [reply libqpdf/QPDF_String.cc]: I've eliminated the need for the `& 0xf0` by static casting and also tightened the call to reserve. Both changes are (very slightly) more efficient.

**Pull request merged**